src/frontend/Semantic_error.ml

```diff
@@ -0,0 +1,552 @@
+           with the correct signature."
+          name
+    | InvalidTruncationCDForCCDF ->
+        Fmt.pf ppf
+          "Truncation is only defined if distribution has _lcdf and _lccdf \
+           functions implemented with appropriate signature."
+    | BreakOutsideLoop ->
+        Fmt.pf ppf "Break statements may only be used in loops."
+    | ContinueOutsideLoop ->
+        Fmt.pf ppf "Continue statements may only be used in loops."
+    | ExpressionReturnOutsideReturningFn ->
+        Fmt.pf ppf
+          "Expression return statements may only be used inside returning \
+           function definitions."
+    | VoidReturnOutsideNonReturningFn ->
+        Fmt.pf ppf
+          "Void return statements may only be used inside non-returning \
+           function definitions."
+    | NonDataVariableSizeDecl ->
+        Fmt.pf ppf
+          "Non-data variables are not allowed in top level size declarations."
+    | NonIntBounds ->
+        Fmt.pf ppf
+          "Bounds of integer variable must be of type int. Found type real."
+    | TransformedParamsInt ->
+        Fmt.pf ppf "(Transformed) Parameters cannot be integers."
+    | MismatchFunDefDecl (name, Some ut) ->
+        Fmt.pf ppf "Function '%s' has already been declared to have type %a."
+          name pp_unsizedtype ut
+    | MismatchFunDefDecl (name, None) ->
+        Fmt.pf ppf
+          "Function '%s' has already been declared but type cannot be \
+           determined."
+          name
+    | FunDeclExists name ->
+        Fmt.pf ppf
+          "Function '%s' has already been declared. A definition is expected."
+          name
+    | FunDeclNoDefn ->
+        Fmt.pf ppf "A function is declared without a corresponding definition."
+    | NonRealProbFunDef ->
+        Fmt.pf ppf
+          "Real return type required for probability functions ending in \
+           _log, _lpdf, _lpmf, _lcdf, or _lccdf."
+    | ProbDensityNonRealVariate (Some ut) ->
+        Fmt.pf ppf
+          "Probability density functions require real variates (first \
+           arguments). Instead found %a."
+          pp_unsizedtype ut
+    | ProbDensityNonRealVariate _ ->
+        Fmt.pf ppf
+          "Probability density functions require real variates (first \
+           arguments)."
+    | ProbMassNonIntVariate (Some ut) ->
+        Fmt.pf ppf
+          "Probability mass functions require integer variates (first \
+           arguments). Instead found %a."
+          pp_unsizedtype ut
+    | ProbMassNonIntVariate _ ->
+        Fmt.pf ppf
+          "Probability mass functions require integer variates (first \
+           arguments)."
+    | DuplicateArgNames ->
+        Fmt.pf ppf "All function arguments must have distinct identifiers."
+    | IncompatibleReturnType ->
+        Fmt.pf ppf
+          "Function bodies must contain a return statement of correct type in \
+           every branch."
+end
+
+type t =
+  | TypeError of location_span * TypeError.t
```

Comment:
This is super slick! :-) I should've done that immediately when I wrote it. The problem was that I had no idea what all the error conditions would be and how they should be organized when I started writing the semantic check.